### PR TITLE
Implement DeviceCode flow in MSAL

### PR DESCRIPTION
**Pull request by MarkZuber**
From previous, abandoned, PR #1263 I've added outstanding comments and current status based on updates made here. 

henrik: not sure this flow should be made available on mobile (as those have browsers)... though not sure I can see a big issue in having it there...
* addressed, it's not in mobile, it's now added via features like username/password

henrik: naming, AquireTokenWithDeviceCodeAsync .... and per other comment, this might have to move to specific platforms.
* addressed: changed name

henrik:   agree on exponential, however I put "exponential" in my question to indicate something different from linear. but it might be fine with something linear for the duration (expected about 15 mins of polling). We should though still ensure that we have to deal with 429's correctly throughout the stack (currently expecting that will be lower in the stack than here).
* added comment in polling loop to address 429 concerns

bogdan: Arg null check in devicecoderequest constructor
* fixed

bogdan: I like the whole callback approach, although it is quite different than what was done in ADAL. A few thoughts:
We should check that the user actually passes in a callback and provide a good error message if he does not
The callback could be async. Sync code is quite frowned upon, especially in ASP.NET scenarios. You can easily wrap sync code in an async callback, but not so much the opposite way around.
* fixed, made this a Func<DeviceCodeResult, Task> to allow it to be await/async.
* as far as the API surface itself, let's finalize discussion with JM next week if we need to modify.

henrik:
Should provide some explanation. Also pls. provide a link to the wiki which talks about how to use the cancellation Token. Also pls. ensure that the usage of Code Flow including the cancelation token is in a sample. Until release you can validate with internal build and publish the sample when the release is out.
aside: Perhaps we should wait adding this everywhere until we are ready to implement it everywhere and then in this change only include changes related to Device Code now?
* TODO:  ADD LINK TO WIKI
* resolved -- only wiring cancellation token to device code API

henrik:
not sure if we should use an overload which calls out the arg specifically.
Aside: When an exception like this bubbles up it would be meaningless to the developer (unless they read our internal code), thus the actual place to do validation is at input time in the public API?
* resovled, moved validation of devicecoderesultcallback to PublicClientApplication boundary

henrik:
thinking, we might want to have a Client exception which is an operation Canceled exception, though not sure about this as it will not be standard (but something for the PM when considering the programming model for the developer)
* OperationCanceledException is the common response in .NET for triggering a cancellationtoken.  I'm going to leave this as is until we have a discussion at the API level with JM

### CODE REVIEW

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeRequest.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeRequest.cs:58]: should IsCancellationRequested be queried after each of the async calls made here? #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeResult.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeResult.cs]: want to just rename it already now? #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeResult.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeResult.cs]: rename already now to scope? #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs]: >  [](start = 42, length = 1)

nit: extra space (is copy pasted below as well) #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs]: >< [](start = 21, length = 1)

should have returns information

(applies more places) #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs]: >nformation. [](start = 83, length = 11)

pls. add an aka.ms link for deviceCode and point it to the wiki and start adding information regarding device code there. The wiki page can be a simple place holder for now but the links and everything can be created and added here.

(applies more places) #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/PublicClientApplication.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Features/DeviceCode/PublicClientApplication.cs]: all should have text. most of it will be what is already available in the interface. #Closed

File: msal/tests/dev apps/DesktopTestApp/MainForm.cs
Comment (henrik-me) [inline msal/tests/dev apps/DesktopTestApp/MainForm.cs:474]: Suggestion: Add a button to cancel the request to ensure we have easy e2e access to validate the functionality

**Comment by henrik-me** [comment]
Looks fantastic. 

Below are not blocking the PR, however are things to complete to be "Done Done".

Testing docs:
Pls. review the current test cases on this and ensure those apply to MSAL (and add the stuff to additionally validate for MSAL such as cancellation).

Automated testing: What's the current test automation coverage? How do we get to 100% (relevant) scenario coverage. E.g. Can we meaningfully test automate this feature? (imo, it's possible, thus also a question of should/risk, etc.)
 #Closed

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs]: >B6SUYU5PL [](start = 34, length = 9)

nit: separate const? #Closed

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs]: >Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId); [](start = 12, length = 62)

you have this test above. did you intend to make this test 2 times? perhaps testing returned scopes? #Closed

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs:170]: suggestion: assert token case before and after state? I suppose we are not caching tokens for Device Code? #Closed

**Review by henrik-me — CHANGES_REQUESTED** [review]
:clock1:

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeRequest.cs
Comment (MarkZuber) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeRequest.cs:58]: Eventually, yes.  Given we talked about not exposing cancellation to anyone else, the main place you'll want to cancel devicecode is in the polling state which happens within only one of these methods.  so I think we should do the "iscancellationrequested" check once we've more fully embraced cancellation in the main methods.
 #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeResult.cs
Comment (MarkZuber) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeResult.cs]: I'll fix this.  I need to debug this case to look at what comes back for it in the /organizations flow or if this is just the /scopes that the user submitted in the first place. #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs
Comment (MarkZuber) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs]: Do you mean extra space after the period?

https://www.quickanddirtytips.com/education/grammar/two-spaces-after-a-period

yes, i'm one of those who was taught to put 2 spaces after a period and it's been cemented into my muscle memory.  I'll fix these. #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs
Comment (MarkZuber) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs]: fixed. #Closed

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (MarkZuber) [reply msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs:170]: We are writing them to the cache, but we don't try to read them from the cache. #Pending

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/PublicClientApplication.cs
Comment (MarkZuber) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/PublicClientApplication.cs]: Shouldn't we be able to use <inheritdoc/> instead of having to copy/paste the text? #Pending

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs
Comment (MarkZuber) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs]: wiki page added at https://aka.ms/msal-device-code-flow.  

will look at other pages and work on filling out the contents there asynchronously from this PR #Closed

**Comment by MarkZuber** [comment]
>           Looks fantastic.
> Below are not blocking the PR, however are things to complete to be "Done Done".
> Testing docs:
> Pls. review the current test cases on this and ensure those apply to MSAL (and add the stuff to additionally validate for MSAL such as cancellation).
> Automated testing: What's the current test automation coverage? How do we get to 100% (relevant) scenario coverage. E.g. Can we meaningfully test automate this feature? (imo, it's possible, thus also a question of should/risk, etc.)

I will add automation coverage into the other repo once this PR is submitted and see what's already there for ADAL to ensure we have coverage.  will also look at test cases to ensure we're all filled in. #Closed

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (MarkZuber) [reply msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs]: Fixed, wired in scopes.  Also added test for cancel. #Closed

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (MarkZuber) [reply msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs]: fixed #Closed

File: msal/tests/dev apps/DesktopTestApp/MainForm.cs
Comment (MarkZuber) [reply msal/tests/dev apps/DesktopTestApp/MainForm.cs:474]: done and tested

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeResult.cs
Comment (MarkZuber) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeResult.cs]: fixed #Closed

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/IPublicClientApplication.cs]: thanks for educating me. fun to learn something new :)

---
In reply to: [220763901](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1275#discussion_r220763901) [](ancestors = 220763901)

File: msal/src/Microsoft.Identity.Client/Features/DeviceCode/PublicClientApplication.cs
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/Features/DeviceCode/PublicClientApplication.cs]: If it works for reference docs, absolutely, makes fine sense to me. I suppose the current approach is an old habit.

---
In reply to: [220765223](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1275#discussion_r220765223) [](ancestors = 220765223)

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs]: >"BAQABAAEA [](start = 28, length = 10)

did you consider making this one and the usercode consts? perhaps even the url?

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (henrik-me) [reply msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs:170]: Kind of makes sense to enable using the RT to for silent calls going forward. Seems like a good scenario to validate in the e2e testing. Device Code would be using the application cache given there is no user involved. If we can't make silent calls based on client_id alone then I suppose token cashing isn't useful.

---
In reply to: [220764186](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1275#discussion_r220764186) [](ancestors = 220764186)

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs:129]: >AccountCacheDictionary [](start = 57, length = 22)

Which account is used? The one who logged in using the returned code?

**Comment by henrik-me** [comment]
few minor comments and expecting a follow up test PR. pls. sync with u/p release before merging to dev (or hide the public API) before committing to dev.

**Review by henrik-me — APPROVED** [review]
:shipit:

File: msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj]: >None [](start = 5, length = 4)

Hmm, is this supposed to be here? I believe the None group controls visibility in VS Solution Explorer only.

**Comment by henrik-me** [comment]
Related GitHub issue: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/551

File: msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj
Comment (MarkZuber) [reply msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj]: nope, it's not.  good catch.  vs is torturing me ;)  #fixed

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (MarkZuber) [reply msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs]: #fixed

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs
Comment (MarkZuber) [reply msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs:129]: should be yes.  will validate/automate this with end to end tests as we discussed in person.

**Pull request merged**
